File: app/src/main/java/com/quarto/QuartoTouchListener.java
package com.quarto;

import android.annotation.SuppressLint;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.Animation;
import android.view.animation.ScaleAnimation;
import android.widget.RelativeLayout;

class QuartoTouchListener implements View.OnTouchListener {

    private int qid;
    private QuartoListener quartoListener;
    private ScaleAnimation scaleIn, scaleOut;
    private float leftMargin, topMargin;

    QuartoTouchListener(int qid, QuartoListener quartoListener) {
        this.qid = qid;
        this.quartoListener = quartoListener;

        scaleIn =  new ScaleAnimation(
                1f,
                1.2f,
                1f,
                1.2f,
                Animation.RELATIVE_TO_SELF,
                0.5f,
                Animation.RELATIVE_TO_SELF,
                0.5f
        );
        scaleIn.setInterpolator(new AccelerateInterpolator());
        scaleIn.setDuration(50);
        scaleIn.setFillAfter(true);

        scaleOut =  new ScaleAnimation(
                1.2f,
                1f,
                1.2f,
                1f,
                Animation.RELATIVE_TO_SELF,
                0.5f,
                Animation.RELATIVE_TO_SELF,
                0.5f
        );
        scaleOut.setInterpolator(new AccelerateInterpolator());
        scaleOut.setDuration(20);
        scaleOut.setFillAfter(true);

    }

    @SuppressLint("ClickableViewAccessibility")
    public boolean onTouch(View view, MotionEvent event) {

        final float rawX = event.getRawX();
        final float rawY = event.getRawY();

        switch (event.getAction() & MotionEvent.ACTION_MASK) {

            case MotionEvent.ACTION_DOWN:
                RelativeLayout.LayoutParams lParams = (RelativeLayout.LayoutParams) view.getLayoutParams();
                leftMargin = rawX - lParams.leftMargin;
                topMargin = rawY - lParams.topMargin;
                view.startAnimation(scaleIn);
                break;

            case MotionEvent.ACTION_UP:
                view.startAnimation(scaleOut);
                quartoListener.onDrop(qid, view);
                break;

            case MotionEvent.ACTION_POINTER_DOWN:
                break;

            case MotionEvent.ACTION_POINTER_UP:
                break;

            case MotionEvent.ACTION_MOVE:
                RelativeLayout.LayoutParams layoutParams = (RelativeLayout.LayoutParams) view.getLayoutParams();
                layoutParams.leftMargin = (int) (rawX - leftMargin);
                layoutParams.topMargin = (int) (rawY - topMargin);
                view.setLayoutParams(layoutParams);
                view.requestLayout();
                break;
        }

        return true;
    }
}

File: app/src/main/java/com/quarto/QuartoTouchListener.kt
package com.quarto

import android.annotation.SuppressLint
import android.os.Handler
import android.util.Log
import android.view.MotionEvent
import android.view.View
import kotlin.math.max
import kotlin.math.min

class QuartoTouchListener(private val qid: Int, private val quartoListener: QuartoListener) : View.OnTouchListener {

    // Coordinates
    private var widgetXFirst: Float = 0F
    private var widgetDX: Float = 0F
    private var widgetYFirst: Float = 0F
    private var widgetDY: Float = 0F
    private var widgetLastAction: Int = 0

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouch(view: View, event: MotionEvent): Boolean {

        if (view.isEnabled) {

            when (event.action and MotionEvent.ACTION_MASK) {

                MotionEvent.ACTION_DOWN -> {

                    view.bringToFront()
                    if (quartoListener.onDown(qid)) {
                        this.widgetDX = view.x - event.rawX
                        this.widgetDY = view.y - event.rawY
                        this.widgetXFirst = view.x
                        this.widgetYFirst = view.y
                        this.widgetLastAction = MotionEvent.ACTION_DOWN
                        //view.animate().scaleXBy(.2f).scaleYBy(.2f).setDuration(200).start()
                    }
                }

                MotionEvent.ACTION_MOVE -> {

                    if (quartoListener.onMove(qid)) {

                        val viewParent: View = (view.parent as View)
                        val parentHeight = viewParent.height
                        val parentWidth = viewParent.width

                        // Screen border Collision
                        var newX = event.rawX + this.widgetDX
                        newX = max(0F, newX)
                        newX = min((parentWidth - view.width).toFloat(), newX)
                        view.x = newX

                        var newY = event.rawY + this.widgetDY
                        newY = max(0F, newY)
                        newY = min((parentHeight - view.height).toFloat(), newY)
                        view.y = newY

                        this.widgetLastAction = MotionEvent.ACTION_MOVE

                        quartoListener.onMoved(parentWidth, parentHeight, newX, newY)
                    }
                }

                MotionEvent.ACTION_UP -> {

                    if (quartoListener.onDrop(qid)) {

                    }
                }

                else -> return false

            }

        }

        return true
    }
}